test/extension/numeric/kernel.cpp

```diff
@@ -34,6 +48,20 @@ BOOST_AUTO_TEST_CASE(kernel_1d_parameterized_constructor)
     BOOST_TEST(k.size() == 9);
 }
 
+BOOST_AUTO_TEST_CASE(kernel_2d_parameterized_constructor)
+{
+    gil::kernel_2d<int> k(9, 4, 4);
+    BOOST_TEST(k.center_vertical() == 4);
+    BOOST_TEST(k.center_horizontal() == 4);
```

Comment:
Given 2D kernel is a rectangular array of numbers, addressing element in the center using 1D indexing seems ambiguous. Especially that no fat comment is provided on that peculiarity and its rationale.

Let's consider canonical form of a 3x3 kernel (symmetric, with center element in the center), then the indexing is as follows (column-major order):

```
| 0,0 | 1,0 | 2,0 |
| 0,1 | 1,1 | 2,1 |
| 0,2 | 1,2 | 2,2 |
```

Thus, most readers, I think, would have expected these:

```cpp
BOOST_TEST(k.center_vertical() == 1);
BOOST_TEST(k.center_horizontal() == 1);
```

Alternative indexing convention often used for the 2D kernel is with center at `0,0` as explained:
 
- http://www.songho.ca/dsp/convolution/convolution2d_example.html
- http://www.songho.ca/dsp/convolution/convolution.html#convolution_2d

```
| -1,-1 | 0,-1 | 1,-1 |
| -1, 0 | 0, 0 | 1, 0 |
| -1, 1 | 0, 1 | 1, 1 |
```

The sizes up/down/left/right do not seem to be correct either.

Reply:
@miralshah365 As I said on Gittter, you've probably read more image processing literature than I have, so you've seen many indexing conventions and should know better which is canonical or common.

However, I can see why some prefer to refer to kernel center at `(0, 0)`, because
- it is intuitive to have center at constant location (for any odd and square kernel)
- it makes arithmetic to access a pixel neighbourhood easy, especially for the optimal linear storage of image data, for example, in `3x3` kernel:

    ```
    // 1 is left/right/up/down size from the center in 3x3 kernel
    pixel_index = (pixel_x + kernel_i - 1) + image_width * (pixel_y + kernel_j - 1);
    ```

One image for thousand words :)

![image](https://user-images.githubusercontent.com/80741/62324158-2b462200-b4a9-11e9-988b-865745ba9f75.png)
(Source: https://www.cs.utexas.edu/~theshark/courses/cs324e/lectures/cs324e-6.pdf)

Obviously, the pattern becomes trickier for even kernels e.g. `4x4`

Reply:
> Let's consider canonical form of a 3x3 kernel (symmetric, with center element in the center), then the indexing is as follows (column-major order):
> 
> | 0,0 | 1,0 | 2,0 |
> | 0,1 | 1,1 | 2,1 |
> | 0,2 | 1,2 | 2,2 |

I chose to go with this style of indexing because it is easier to maintain and indexing will behave the same as any 2D array-like and I believe this will be less confusing for the user instead of remembering the different style of indexing.